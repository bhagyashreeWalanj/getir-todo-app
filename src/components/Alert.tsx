import React, { useEffect } from 'react'
import styles from '../styles/alert.module.scss'
import { default as classnames } from 'classnames'
import { CgDanger } from 'react-icons/cg'
import { TiTick } from 'react-icons/ti'

export interface IToDoAlert {
  type: string
  msg: string
  removeAlert: () => void
}

const Alert = ({ type, msg, removeAlert }: IToDoAlert) => {
  useEffect(() => {
    const timeout = setTimeout(() => {
      removeAlert()
    }, 3000)
    return () => clearTimeout(timeout)
  })

  return (
    <div
      className={classnames(
        type === 'success-edit'
          ? [styles.editSuccessAlert]
          : type === 'danger-edit'
          ? [styles.editErrorAlert]
          : [styles.alert],
      )}
    >
      <p
        className={classnames({
          [styles.errormsg]: type === 'danger',
          [styles.successmsg]: type === 'success',
          [styles.errormsgEdit]: type === 'danger-edit',
          [styles.successmsgEdit]: type === 'success-edit',
        })}
        id={`alertDiv`}
      >
        {type === 'danger' ? (
          <CgDanger className={styles.dangerIcon} />
        ) : (
          <TiTick />
        )}{' '}
        {msg}
      </p>
    </div>
  )
}

export default Alert
