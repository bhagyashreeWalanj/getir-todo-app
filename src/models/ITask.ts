export interface ITask {
    id: string;
    title: string;
    createdAt: string;
    completed: boolean;
  }
  