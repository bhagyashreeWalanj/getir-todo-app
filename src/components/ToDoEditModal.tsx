import React, { useState } from 'react'
import Modal from 'react-modal'
import styles from '../styles/modal.module.scss'
import { RiCloseCircleFill } from 'react-icons/ri'
import Alert from './Alert'

interface IEditTaskModalProps {
  taskName: string
  isOpen: boolean
  onRequestClose: () => void
  handleEditTask: (id: string, title: string) => void
  EditTaskId: string
}

const ToDoEditModal = ({
  taskName,
  isOpen,
  onRequestClose,
  handleEditTask,
  EditTaskId,
}: IEditTaskModalProps) => {
  const [newTitle, setNewTitle] = useState('')
  const [alert, setAlert] = useState({ show: false, msg: '', type: '' })

  const showAlert = (show = false, type = '', msg = '') => {
    setAlert({ show, type, msg })
  }

  const handleEditOnChange = (title: string) => {
    setNewTitle(title)
  }
  const editTask = (id: string, title: string) => {
    if (title !== '') {
      handleEditTask(id, title)
      showAlert(true, 'success-edit', 'Successfully changed')
      const timeout = setTimeout(() => {
        onRequestClose()
        setAlert({ show: false, msg: '', type: '' })
      }, 2000)
      return () => clearTimeout(timeout)
    } else {
      showAlert(true, 'danger-edit', 'Please add task to save')
      return
    }
  }

  return (
    <div>
      <Modal
        isOpen={isOpen}
        onRequestClose={onRequestClose}
        className={styles.modalContainer}
        overlayClassName={styles.overlayContainer}
        ariaHideApp={false}
      >
        <RiCloseCircleFill
          className={styles.closeButton}
          onClick={() => {
            onRequestClose()
          }}
        />
        {alert.show && <Alert {...alert} removeAlert={showAlert} list={[]} />}

        {!!taskName && (
          <div key={EditTaskId}>
            <div className={styles.label}>
              <h2 className={styles.header}>Edit Task</h2>
            </div>

            <div className={styles.NoteTypeContainer}>
              <input
                type="text"
                className={styles.inputText}
                placeholder="Edit task..."
                //value={task}
                defaultValue={taskName}
                name={`input_${EditTaskId}`}
                onChange={(e) => handleEditOnChange(e.target.value)}
              />

              <button
                className={styles.ButtonSave}
                onClick={() => {
                  editTask(EditTaskId, newTitle)
                }}
              >
                <span>Save</span>
              </button>
            </div>
          </div>
        )}
      </Modal>
    </div>
  )
}

export default ToDoEditModal
