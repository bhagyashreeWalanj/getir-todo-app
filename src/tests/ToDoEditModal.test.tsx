import React from 'react'
import { shallow, mount } from 'enzyme'
import ToDoAppList from '../components/ToDoList'
import { ITask } from '../models/ITask'
import ToDoEditModal from '../components/ToDoEditModal'
import Alert from '../components/Alert'

describe('Testing ToDo Edit Modal Component', () => {
  const editTask = jest.fn()
  const removeTask = jest.fn()
  const taskList: ITask[] = [
    {
      id: '1',
      title: 'Hello World 1',
      createdAt: '01/05/2022',
      completed: false,
    },
    {
      id: '2',
      title: 'Hello World 2',
      createdAt: '01/05/2022',
      completed: false,
    },
  ]

  const wrapperEdit = mount(
    <ToDoAppList
      tasks={taskList}
      editTask={editTask}
      removeTask={removeTask}
    />,
  )
  wrapperEdit.find('FiEdit#editBtn_2').simulate('click', {
    id: '2',
  })

  const handleEditTask = jest.fn()
  const onRequestClose = jest.fn()
  const modalWrapper = mount(
    <ToDoEditModal
      isOpen={true}
      EditTaskId={'2'}
      taskName={taskList[1].title}
      handleEditTask={handleEditTask}
      onRequestClose={onRequestClose}
    />,
  )

  it('should display to Edit task Modal Component', () => {
    expect(modalWrapper).toMatchSnapshot()
  })

  it('should call edit modal on click edit Button', () => {
    expect(wrapperEdit.find(ToDoEditModal)).toHaveLength(1)
  })

  it('should display selected Task title on Edit Modal Input', () => {
    expect(modalWrapper.find('.inputText').props().defaultValue).toBe(
      'Hello World 2',
    )
  })

  it('should show error if input is empty', () => {
    modalWrapper
      .find('.inputText')
      .simulate('change', { target: { value: '' } })
    modalWrapper.find('.ButtonSave').simulate('click')

    expect(modalWrapper.find(Alert)).toHaveLength(1)
    expect(modalWrapper.find(Alert).find('#alertDiv').text()).toContain(
      'Please add task to save',
    )
  })

  it('should modal close after successful edit action', () => {
    modalWrapper
      .find('.inputText')
      .simulate('change', { target: { value: 'Play Basketball' } })
    modalWrapper.find('.ButtonSave').simulate('click')

    expect(modalWrapper.find(Alert).props().msg).toBe('Successfully changed')
    setTimeout(() => {
      expect(onRequestClose).toBeCalledTimes(1)
    }, 3000)
  })
})
