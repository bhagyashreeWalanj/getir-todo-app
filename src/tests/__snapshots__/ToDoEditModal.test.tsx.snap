// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ToDo Edit Modal Component should display to Edit task Modal Component 1`] = `
<ToDoEditModal
  EditTaskId="2"
  handleEditTask={[MockFunction]}
  isOpen={true}
  onRequestClose={[MockFunction]}
  taskName="Hello World 2"
>
  <div>
    <Modal
      ariaHideApp={false}
      bodyOpenClassName="ReactModal__Body--open"
      className="modalContainer"
      closeTimeoutMS={0}
      contentElement={[Function]}
      isOpen={true}
      onRequestClose={[MockFunction]}
      overlayClassName="overlayContainer"
      overlayElement={[Function]}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      preventScroll={false}
      role="dialog"
      shouldCloseOnEsc={true}
      shouldCloseOnOverlayClick={true}
      shouldFocusAfterRender={true}
      shouldReturnFocusAfterClose={true}
    >
      <Portal
        containerInfo={
          <div
            class="ReactModalPortal"
          >
            <div
              class="ReactModal__Overlay overlayContainer"
            >
              <div
                aria-modal="true"
                class="ReactModal__Content modalContainer"
                role="dialog"
                tabindex="-1"
              >
                <svg
                  class="closeButton"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g>
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm0-11.414L9.172 7.757 7.757 9.172 10.586 12l-2.829 2.828 1.415 1.415L12 13.414l2.828 2.829 1.415-1.415L13.414 12l2.829-2.828-1.415-1.415L12 10.586z"
                    />
                  </g>
                </svg>
                <div>
                  <div
                    class="label"
                  >
                    <h2
                      class="header"
                    >
                      Edit Task
                    </h2>
                  </div>
                  <div
                    class="NoteTypeContainer"
                  >
                    <input
                      class="inputText"
                      name="input_2"
                      placeholder="Edit task..."
                      type="text"
                      value="Hello World 2"
                    />
                    <button
                      class="ButtonSave"
                    >
                      <span>
                        Save
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        }
      >
        <ModalPortal
          ariaHideApp={false}
          bodyOpenClassName="ReactModal__Body--open"
          className="modalContainer"
          closeTimeoutMS={0}
          contentElement={[Function]}
          defaultStyles={
            Object {
              "content": Object {
                "WebkitOverflowScrolling": "touch",
                "background": "#fff",
                "border": "1px solid #ccc",
                "borderRadius": "4px",
                "bottom": "40px",
                "left": "40px",
                "outline": "none",
                "overflow": "auto",
                "padding": "20px",
                "position": "absolute",
                "right": "40px",
                "top": "40px",
              },
              "overlay": Object {
                "backgroundColor": "rgba(255, 255, 255, 0.75)",
                "bottom": 0,
                "left": 0,
                "position": "fixed",
                "right": 0,
                "top": 0,
              },
            }
          }
          isOpen={true}
          onRequestClose={[MockFunction]}
          overlayClassName="overlayContainer"
          overlayElement={[Function]}
          parentSelector={[Function]}
          portalClassName="ReactModalPortal"
          preventScroll={false}
          role="dialog"
          shouldCloseOnEsc={true}
          shouldCloseOnOverlayClick={true}
          shouldFocusAfterRender={true}
          shouldReturnFocusAfterClose={true}
          style={
            Object {
              "content": Object {},
              "overlay": Object {},
            }
          }
        >
          <div
            className="ReactModal__Overlay overlayContainer"
            onClick={[Function]}
            onMouseDown={[Function]}
            style={Object {}}
          >
            <div
              aria-modal={true}
              className="ReactModal__Content modalContainer"
              onClick={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              role="dialog"
              style={Object {}}
              tabIndex="-1"
            >
              <RiCloseCircleFill
                className="closeButton"
                onClick={[Function]}
              >
                <IconBase
                  attr={
                    Object {
                      "viewBox": "0 0 24 24",
                    }
                  }
                  className="closeButton"
                  onClick={[Function]}
                >
                  <svg
                    className="closeButton"
                    fill="currentColor"
                    height="1em"
                    onClick={[Function]}
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g
                      key="0"
                    >
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                        key="0"
                      />
                      <path
                        d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm0-11.414L9.172 7.757 7.757 9.172 10.586 12l-2.829 2.828 1.415 1.415L12 13.414l2.828 2.829 1.415-1.415L13.414 12l2.829-2.828-1.415-1.415L12 10.586z"
                        key="1"
                      />
                    </g>
                  </svg>
                </IconBase>
              </RiCloseCircleFill>
              <div
                key="2"
              >
                <div
                  className="label"
                >
                  <h2
                    className="header"
                  >
                    Edit Task
                  </h2>
                </div>
                <div
                  className="NoteTypeContainer"
                >
                  <input
                    className="inputText"
                    defaultValue="Hello World 2"
                    name="input_2"
                    onChange={[Function]}
                    placeholder="Edit task..."
                    type="text"
                  />
                  <button
                    className="ButtonSave"
                    onClick={[Function]}
                  >
                    <span>
                      Save
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </ModalPortal>
      </Portal>
    </Modal>
  </div>
</ToDoEditModal>
`;
